src/riak_kv_qry_worker.erl

```diff
@@ -261,43 +262,54 @@ throttling_spawn_index_fsms(State) ->
 estimate_query_size(#state{n_subqueries_done = NSubqueriesDone} = State)
   when NSubqueriesDone < 2  ->
     State;
-estimate_query_size(#state{total_query_data  = CurrentTotalSize,
+
+estimate_query_size(#state{total_query_data  = TotalQueryData,
+                           total_query_rows  = TotalQueryRows,
                            n_subqueries_done = NSubqueriesDone,
                            max_query_data    = MaxQueryData,
                            qbuf_ref          = QBufRef,
                            sub_qrys          = SubQrys,
                            qry = ?SQL_SELECT{'LIMIT' = [Limit]} = OrigQry} = State)
-  when QBufRef /= undefined,
-       is_integer(Limit) ->
+  when QBufRef /= undefined ->
+
     %% query buffer-backed, has a LIMIT: consider the latter
-    BytesPerChunk = CurrentTotalSize / NSubqueriesDone,
-    ProjectedLimitData = round(Limit * BytesPerChunk),
-    if ProjectedLimitData > MaxQueryData ->
-            lager:info("Cancelling LIMIT ~b query because projected result size exceeds limit (~b > ~b, subqueries ~b of ~b done, query ~p)",
-                       [Limit, ProjectedLimitData, MaxQueryData, NSubqueriesDone, length(SubQrys), OrigQry]),
+    EstLimitData = round(Limit * (TotalQueryData / TotalQueryRows)),
+    IsLimitTooBig = EstLimitData > MaxQueryData,
+
+    %% but also check the grand total, for the case when LIMIT is big
+    %% but WHERE range is still tiny
+    EstTotalData = round(TotalQueryData + (TotalQueryData / NSubqueriesDone) * length(SubQrys)),
+    IsWhereTooBig = EstTotalData > MaxQueryData,
+
+    case IsLimitTooBig and IsWhereTooBig of
```

Comment:
Thank you. The match in logic between the code and log message clearly means that we are being conservative about cancelling queries since both conditions must be true.